### PR TITLE
Update to .NET 9

### LINKED ISSUE #2669: Update to .Net 9
The SqlClient build system needs to be updated to dotnet 9 to be able to run the CI and test SqlDbType.Json added in dotnet 9.

### PR DESCRIPTION
**Pull request by ErikEJ**
fixes #2669

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -2,6 +2,6 @@
 <configuration>
   <packageSources>
     <clear />
-    <add key="sqlclient" value="https://sqlclientdrivers.pkgs.visualstudio.com/public/_packaging/sqlclient/nuget/v3/index.json" />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
```
Comment (ErikEJ) [inline NuGet.config:5]: Need the new packages in the sqlclient feed or a new feed

File: eng/pipelines/libraries/mds-validation-variables.yml [hunk 2]
```diff
@@ -15,15 +15,15 @@ variables:
...
+    value: net462,net6.0,net8.0,net9.0
   - name: Database
     value: Northwind
   - name: platform
     value: AnyCPU
   - name: TargetNetFxVersion
     value: net481
   - name: TargetNetCoreVersion
-    value: net8.0
+    value: net9.0
```
Comment (ErikEJ) [inline eng/pipelines/libraries/mds-validation-variables.yml:26]: Not sure what the exact purpose of this value is

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlNormalizer.cs [hunk 3]
```diff
@@ -152,7 +152,11 @@ internal sealed class BinaryOrderedUdtNormalizer : Normalizer
                 if (nullByte == 0)
                 {
                     result = _nullInstance;
-                    s.Read(_padBuffer, 0, _padBuffer.Length);
+#if NET8_0_OR_GREATER
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlNormalizer.cs:155]: Fixes the new CA2022

File: src/Microsoft.Data.SqlClient/tests/Directory.Build.props [hunk 4]
```diff
@@ -12,6 +12,7 @@
     <Configurations>Debug;Release;</Configurations>
     <Platforms>AnyCPU;x86;x64</Platforms>
     <ReferenceType Condition="'$(ReferenceType)'==''">Project</ReferenceType>
+    <NuGetAudit>false</NuGetAudit>
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/tests/Directory.Build.props:15]: To prevent NuGet audit on test projects giving warnings as errors on old packages

**Comment by DavoudEshtehari** [comment]
FYI, the pipeline is expected to fail because lack of `.NET` 9 on running agents.

**Comment by ErikEJ** [comment]
@DavoudEshtehari Yes, that was expected, but it did not even fail ??

**Comment by ErikEJ** [comment]
@JRahnama Thanks, much better error this time 😄 

`The current .NET SDK does not support targeting .NET 9.0.`

**Comment by saurabh500** [comment]
@cheenamalhotra @DavoudEshtehari @JRahnama @David-Engel  What do we need to do to move this along? 
Our build agents require .Net runtime 9 installed. Is the runtime supposed to be baked into the image itself? Or is there a script which downloads and installs the runtime?

**Comment by JRahnama** [comment]
> @cheenamalhotra @DavoudEshtehari @JRahnama @David-Engel What do we need to do to move this along? Our build agents require .Net runtime 9 installed. Is the runtime supposed to be baked into the image itself? Or is there a script which downloads and installs the runtime?

1ES images do not have Net9.0 added. You can see included SDK in 1ES at [here](https://github.com/actions/runner-images/blob/win22/20240707.1/images/windows/Windows2022-Readme.md#net-core-tools).

**Comment by DavoudEshtehari** [comment]
> @cheenamalhotra @DavoudEshtehari @JRahnama @David-Engel What do we need to do to move this along? Our build agents require .Net runtime 9 installed. Is the runtime supposed to be baked into the image itself? Or is there a script which downloads and installs the runtime?

To streamline the process and follow the pattern, a new artifact should be included in the images to reduce installation time for each run. But the compatible VS would be required though.

**Comment by ErikEJ** [comment]
@JRahnama  I just use this setting in VS Options:

image

**Comment by JRahnama** [comment]
> @JRahnama I just use this setting in VS Options:
> 
> image

but if you look into VS Installer you probably will see something similar as below. 
However, I will try running them with a test pipeline. We may end up getting good results with C# 13 and not having VS 2022-preview.
image

**Comment by JRahnama** [comment]
@ErikEJ , hope you don't mind me pushing the yaml changes into your PR.

File: NuGet.config (hunk 1 above)
Comment (David-Engel) [reply NuGet.config:5]: The sqlclientdrivers feed upstreams from NuGet.org. If you revert this and push, I believe the CI will have permissions to pull from upstream and install the new versions to the feed.

Comment (ErikEJ) [reply NuGet.config:5]: @David-Engel It did not work for me locally, but maybe things have changed?

Comment (ErikEJ) [reply NuGet.config:5]: Got it! I will revert

**Comment by David-Engel** [comment]
I think this will need a change in ci-prebuild-step.yml like:

```
- task: UseDotNet@2
  inputs:
    packageType: 'sdk'
    version: '9.x'
    includePreviewVersions: true
```

**Comment by arellegue** [comment]
When tested on a local machine having .Net9.0 preview, Code [SYSLIB0057](https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0057) was thrown by the compiler for the following files:

- SNICommon,cs
- VirtualSecureModeEnclaveProviderBase.cs

**Comment by JRahnama** [comment]
> When tested on a local machine having .Net9.0 preview, Code [SYSLIB0057](https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0057) was thrown by the compiler for the following files:
> 
> * SNICommon,cs
> * VirtualSecureModeEnclaveProviderBase.cs

Basically you need to change the code to something similar to below:
in SNICommon:

```C#
#if !NET9_0_OR_GREATER
                        validationCertificate = new X509Certificate(validationCertFileName);
#else
                        validationCertificate = X509CertificateLoader.LoadCertificateFromFile(validationCertFileName);
#endif
```

in VirtualSecureModeEnclaveProviderBase.cs:
```C#
#if !NET9_0_OR_GREATER
                    certificateCollection.Import(data);
#else
                   _ =  X509CertificateLoader.LoadCertificate(data);
#endif
```

and in VirtualSecureModeEnclaveProvider.cs:
```C#
#if !NET9_0_OR_GREATER
            Certificate = new X509Certificate2(payload);
#else
            _ = X509CertificateLoader.LoadCertificate(payload);
#endif
```

After that you will get 5 more errors such as below:
```MSBuild
SqlClient\artifacts\Project\obj\Debug.AnyCPU\Microsoft.Data.SqlClient\netcore\net9.0\Microsoft.Data.SqlClie
nt.notsupported.cs(798,37): error IL2093: 'DynamicallyAccessedMemberTypes' in 'DynamicallyAccessedMembersAttribute' on
the return value of method 'Microsoft.Data.SqlClient.SqlDataReader.GetFieldType(Int32)' don't match overridden return v
alue of method 'System.Data.IDataRecord.GetFieldType(Int32)'. All overridden members must have the same 'DynamicallyAcc
essedMembersAttribute'
```

Now we have to address those.

**Comment by arellegue** [comment]
For the new CA2022 error, in addition to #if NET8_0_OR_GREATER for ReadExactly, since the compiler is throwing error for TargetFramework=net6.0, a #pragma warning disable/restore CA2022 was added for the #else block.

**Comment by arellegue** [comment]
@ErikEJ, can I push minor fixes into this branch for issues found when compiling MDS locally?

**Comment by JRahnama** [comment]
> For the new CA2022 error, in addition to #if NET8_0_OR_GREATER for ReadExactly, since the compiler is throwing error for TargetFramework=net6.0, a #pragma warning disable/restore CA2022 was added for the #else block.

Preferably, add them the .editorconfig file inside netcore folder

**Comment by ErikEJ** [comment]
@arellegue go ahead...

**Comment by arellegue** [comment]
A "dotnet_diagnostic.CA2022.severity = error" was added to .editorconfig instead since it is the preferred method.

**Comment by cheenamalhotra** [comment]
Hi @ErikEJ 

I will proceed to close the PR, as PR https://github.com/dotnet/SqlClient/pull/2946 is actively being worked upon and will replace this PR once images are available with latest .NET 9 SDK

**Pull request closed without merging**